src/py/crankshaft/crankshaft/segmentation/segmentation.py

```diff
@@ -0,0 +1,148 @@
+
+    return replace_nan_with_mean(target), replace_nan_with_mean(features)
+
+# High level interface
+# --------------------
+
+def create_and_predict_segment_agg(target, features, target_features, target_ids, model_parameters):
+    """
+
+    """
+
+    clean_target = replace_nan_with_mean(target)
+    clean_features = replace_nan_with_mean(features)
+    target_features = replace_nan_with_mean(target_features)
+
+    model, accuracy = train_model(clean_target,clean_features, model_parameters, 0.2)
+    prediction = model.predict(target_features)
+    return zip(target_ids, prediction, np.full(prediction.shape, accuracy))
+
+
+
+def create_and_predict_segment(query, variable, target_query, model_params):
+    """
+    generate a segment with machine learning
+    Stuart Lynn
+    """
+
+    columns = plpy.execute('SELECT * FROM ({query}) As a LIMIT 1  '.format(query=query))[0].keys()
+
+    feature_columns = set(columns) - set([variable, 'cartodb_id', 'the_geom', 'the_geom_webmercator'])
+    target,features = get_data(variable, feature_columns, query)
+
+    model, accuracy = train_model(target,features, model_params, 0.2)
+    cartodb_ids, result = predict_segment(model,feature_columns,target_query)
+    return zip(cartodb_ids, result, np.full(result.shape, accuracy ))
+
+
+def train_model(target, features, model_params, test_split):
+    """
+
+    """
+    features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=test_split)
+    model = GradientBoostingRegressor(**model_params)
+    model.fit(features_train, target_train)
+    accuracy = calculate_model_accuracy(model,features,target)
+    return model, accuracy
+
+def calculate_model_accuracy(model, features, target):
+    """
+        Calculate the mean squared error of the model prediction
+        Input:
+            @param model: model trained from input features
+            @param features: features to make a prediction from
+            @param target: target to compare prediction to
+        Output:
+            mean squared error of the model prection compared to the target
+    """
+    prediction = model.predict(features)
+    return metrics.mean_squared_error(prediction, target)
+
+def predict_segment(model, features, target_query):
+    """
+    predict a segment with machine learning
+    Stuart Lynn
+
+    description of params?
+    """
+
+    batch_size = 1000
+    joined_features = ','.join(['"{0}"::numeric'.format(a) for a in features])
+
+    cursor = plpy.cursor('SELECT Array[{joined_features}] As features FROM ({target_query}) As a'.format(
+        joined_features=joined_features,
+        target_query= target_query))
+
+    results = []
+
+    while True:
+        rows = cursor.fetch(batch_size)
+        if not rows:
+            break
+        batch = np.row_stack([np.array(row['features'], dtype=float) for row in rows])
+
+        #Need to fix this. Should be global mean. This will cause weird effects
```

Comment:
@stuartlynn, what about a `plpy.execute('select avg(col1) a1, avg(col2) a2 from table` before the loop, and then replace the None values in each column with those calculated avgs later in the loop?